libnet/common.c

```diff
@@ -0,0 +1,380 @@
+#include "common.h"
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <fdebug.h>
+#include <limits.h>
+#include <pthread.h>
+#include <semaphore.h>
+#include <signal.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define CLIENT_MAX_RETRIES 30
+
+//bool enqueue_message(tlv *message);
+//bool notify_tag(unsigned char tag);
+
+tlv *message_queue[MAX_MESSAGE_QUEUE];
+sem_t free_message_slots;
+
+tlv** get_message_queue() {
+	return message_queue;
+}
+
+available_tag_t *available_tags;
+size_t available_tags_number;
+
+available_tag_t* get_available_tags_struct(unsigned char tag) {
+	for(unsigned i = 0; i < available_tags_number; i++) {
+		if(available_tags[i].tag == tag)
+			return available_tags+i;
+	}
+	return 0;
+}
+
+bool libnet_init(unsigned char *tags_to_register, unsigned tags_to_register_number) {
+	available_tags = malloc(tags_to_register_number * sizeof * available_tags);
+	check_mem(available_tags);
+
+	for(unsigned i = 0; i < tags_to_register_number; i++) {
+		available_tags[i].tag = tags_to_register[i];
+		check1(!sem_init(&available_tags[i].number, 0, 0), "tags sem_init");
+	}
+
+	available_tags_number = tags_to_register_number;
+
+	check1(!sem_init(&free_message_slots, 0, MAX_MESSAGE_QUEUE), "message_queue sem_init");
+
+	return true;
+error:
+	if(available_tags)
+		free(available_tags);
+	for(unsigned i = 0; i < tags_to_register_number; i++) {
+		log_info("Cleaning up %u semapthores, this may produce errors",
+				tags_to_register_number);
+		check_warn(sem_destroy(&available_tags[i].number), "tags sem_destroy %u", i);
+	}
+
+	return false;
+}
+
+
+bool enqueue_message(tlv *message) {
+	check1(sem_wait(&free_message_slots), "wait free_message_slots");
+
+	unsigned empty_slot_index;
+	for(empty_slot_index = 0;
+		empty_slot_index < MAX_MESSAGE_QUEUE && message_queue[empty_slot_index];
+		empty_slot_index++);
+
+	check1(empty_slot_index < MAX_MESSAGE_QUEUE,
+			"message_queue in inconsistent state!") ;
+
+	message_queue[empty_slot_index] = message;
+
+	return true;
+
+error:
+	// TODO(florek) error handling
+	return false;
+}
+
+
+static void handle_echo_message(client_info *client, tlv *message, bool reply) {
+	UNUSED(message);
+	debug1("Handle echo");
+	client->retries = CLIENT_MAX_RETRIES;
+	if(reply) {
+		check1(send_tag(client, TAG_OK, 0, 0), "Echo response");
```

Comment:
variable `reply` is used here in meaning `should_i_reply`. 
Guess it could be more clear, will fix.